admin-dev/themes/new-theme/js/pages/product/edit/attachments-manager.js

```diff
@@ -0,0 +1,167 @@
+    this.initAddAttachmentIframe();
+    this.$attachmentsContainer.on('click', ProductMap.attachments.removeAttachmentBtn, (e) => {
+      this.removeAttachmentRow(e);
+    });
+  }
+
+  /**
+   * @private
+   */
+  initAddAttachmentIframe() {
+    this.$attachmentsContainer.find(ProductMap.attachments.addAttachmentBtn).fancybox({
+      type: 'iframe',
+      width: '90%',
+      height: '90%',
+      afterLoad: () => {
+        const form = $('.fancybox-iframe').contents().find('form');
+        const {fancybox} = parent.$;
+
+        form.on('click', '.cancel-btn', () => {
+          fancybox.close();
+        });
+
+        form.submit((e) => {
+          this.eventEmitter.emit(ProductEventMap.attachments.newAttachmentSubmitted, {e, fancybox});
+        });
+      },
+    });
+  }
+
+  /**
+   * @private
+   */
+  onAttachmentSubmit() {
+    this.eventEmitter.on(ProductEventMap.attachments.newAttachmentSubmitted, (eventData) => {
+      const submitEvent = eventData.e;
+      submitEvent.preventDefault();
+
+      createAttachment(submitEvent.currentTarget).then((resp) => {
+        if (resp.errors) {
+          Object.values(resp.errors).forEach((error) => {
+            $.growl.error({message: error});
+          });
+
+          return;
+        }
+        eventData.fancybox.close();
+
+        getAttachmentInfo(resp.attachmentId).then((response) => {
+          this.addAttachmentRow(response.attachmentInfo);
+        });
+      });
+    });
+  }
+
+  /**
+   * @param {Object} event
+   *
+   * @private
+   */
+  removeAttachmentRow(event) {
+    const $removeButton = $(event.currentTarget);
+    const $thisRow = $removeButton.closest(ProductMap.attachments.attachedFileRow);
+    const attachmentId = $thisRow.find(ProductMap.attachments.attachmentIdInputs).val();
+
+    $thisRow.remove();
+    this.eventEmitter.emit(ProductEventMap.attachments.rowRemoved, {attachmentId});
+    this.refreshState();
+  }
+
+  /**
+   * @private
+   */
+  addAttachmentRow(attachmentInfo) {
+    const rowIndex = this.$attachmentsTableBody.find(ProductMap.attachments.attachedFileRow).length;
+    const $row = $(this.getPrototypeRow(rowIndex));
+    const attachmentId = attachmentInfo.id;
+    const $attachmentIdInput = $(ProductMap.attachments.tableRow.attachmentIdInput(rowIndex), $row);
+    const $attachmentNamePreview = $(ProductMap.attachments.tableRow.attachmentNamePreview, $row);
+    const $attachmentFilenamePreview = $(ProductMap.attachments.tableRow.attachmentFilenamePreview, $row);
+    const $attachmentTypePreview = $(ProductMap.attachments.tableRow.attachmentTypePreview, $row);
+
+    $attachmentIdInput.val(attachmentId);
+    $attachmentNamePreview.html(attachmentInfo.name);
+    $attachmentFilenamePreview.html(attachmentInfo.filename);
+    $attachmentTypePreview.html(attachmentInfo.mimeType);
+
+    this.$attachmentsTableBody.append($row);
+    this.eventEmitter.emit(ProductEventMap.attachments.rowAdded, {attachmentId});
```

Comment:
are you sure? I find it useful - it serves its purpose for "save" button state, but also it would allow module developers to listen for changes in certain block :man_shrugging:

But if you don't want to care too much about extensibility concerns, then I guess you are right. Just confirm this please :wink:

Reply:
I see what you mean and the purpose And I agree it's interesting, that said in theory this should all be handled via the `FormObjectMapper` js component, it's just that we didn't configure the mapping for the whole page

Which makes me think it's a good work I should start while we are blocked by QA bottleneck, the component already handles some generic event along with individual event and a watcher system So I guess multiplying the events will become redundant once it's correctly and fully configured